pkg/virt-controller/watch/workload-updater/workload-updater.go

```diff
@@ -0,0 +1,554 @@
+	// there can only be a single one of these in a cluster at a time.
+	threadiness := 1
+
+	// Wait for cache sync before we start the controller
+	cache.WaitForCacheSync(stopCh, c.migrationInformer.HasSynced, c.vmiInformer.HasSynced, c.kubeVirtInformer.HasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping workload update controller.")
+}
+
+func (c *WorkloadUpdateController) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *WorkloadUpdateController) Execute() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	err := c.execute(key.(string))
+
+	if err != nil {
+		log.Log.Reason(err).Infof("reenqueuing workload updates for KubeVirt %v", key)
+		c.queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed workload updates for KubeVirt %v", key)
+		c.queue.Forget(key)
+	}
+	return true
+}
+
+func (c *WorkloadUpdateController) isOutdated(vmi *virtv1.VirtualMachineInstance) bool {
+	if vmi.IsFinal() {
+		return false
+	}
+
+	// if the launcher image isn't detected yet, that means
+	// we don't know what the launcher image is yet.
+	// This could be due to a migration, or the VMI is still
+	// initializing. virt-controller will set it for us once
+	// either the VMI is either running or done migrating.
+	if vmi.Labels == nil {
+		return false
+	}
+
+	_, labelExists := vmi.Labels[virtv1.OutdatedLauncherImageLabel]
+	if labelExists {
+		return true
+	}
+
+	return false
+}
+
+func isMigratable(vmi *virtv1.VirtualMachineInstance) bool {
+	for _, c := range vmi.Status.Conditions {
+		if c.Type == virtv1.VirtualMachineInstanceIsMigratable && c.Status == k8sv1.ConditionTrue {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (c *WorkloadUpdateController) getUpdateData(kv *virtv1.KubeVirt) *updateData {
+	data := &updateData{}
+
+	lookup := make(map[string]bool)
+
+	migrations := migrationutils.ListUnfinishedMigrations(c.migrationInformer)
+
+	for _, migration := range migrations {
+		lookup[migration.Namespace+"/"+migration.Spec.VMIName] = true
+	}
+
+	automatedMigrationAllowed := false
+	automatedShutdownAllowed := false
+
+	for _, method := range kv.Spec.WorkloadUpdateStrategy.WorkloadUpdateMethods {
+		if method == virtv1.WorkloadUpdateMethodLiveMigrate {
+			automatedMigrationAllowed = true
+		} else if method == virtv1.WorkloadUpdateMethodShutdown {
+			automatedShutdownAllowed = true
+		}
+	}
+
+	data.numActiveMigrations = len(migrations)
+
+	objs := c.vmiInformer.GetStore().List()
+	for _, obj := range objs {
+		vmi := obj.(*virtv1.VirtualMachineInstance)
+		if !vmi.IsRunning() || vmi.IsFinal() || vmi.DeletionTimestamp != nil {
+			// only consider running VMIs that aren't being shutdown
+			continue
+		} else if !c.isOutdated(vmi) {
+			continue
+		}
+
+		// add label to outdated vmis for sorting
+		data.allOutdatedVMIs = append(data.allOutdatedVMIs, vmi)
```

Comment:
This line does not seem to add a label or do anything related.

Reply:
comment removed